File: 24_SwapNodesinPairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
//解法1
const swapPairs = function(head) {
    
    //若是head === null or head.next === null  就可直接回傳;
    if(head === null || head.next === null) return head;

    //用number紀錄當前loop到第幾個節點
    let number = 1;
    //紀錄head是哪一個node
    let header = head;
    //紀錄當前loop到的節點  起點為head
    let currentNode = head;
    //紀錄loop時的前一個節點
    let previousNode = null;

};

function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

//swapPairs( new ListNode(1,new ListNode(2,new ListNode(3, new ListNode(4, undefined)))))

File: a24.SwapNodesinPairs/24_SwapNodesinPairs.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
#解法1
def swapPairs(head: ListNode) -> ListNode:
    #若head == None  or head.next == None   return head
    if head == None or head.next == None:
        return head
    
    #將head前套上node
    fakeNode = ListNode()
    fakeNode.next = head
    #保存fakeNode地址
    result = fakeNode

    #需要檢查下個及下下個皆有值時才進行調換
    while fakeNode.next != None and fakeNode.next.next != None:
        #先保存下個及下下個的位址
        a = fakeNode.next
        b = fakeNode.next.next

        #改變fakeNode.next為b 0 -> 2
        fakeNode.next = b
        #改變 下個.next = b.next 0 -> 2 -> 1 -> 3
        a.next = b.next
        #改變下下個.next = a  0 -> 2 -> 1
        b.next = a
        #移動fakeNode指針至第二個 0 -> 2 -> 1(指針位置) -> 3
        fakeNode = a

    return result.next


swapPairs(ListNode(1, ListNode(2)))

        
    